### PR TITLE
feat: Support for "Instance Refresh"

### PR DESCRIPTION
**Pull request by spatel96**
## Description
Adding the instance refresh block to the ASG, for now this is a resource that is always created and uses defaults defined in variables. 

Not sure if it would make sense to attempt to place this in a dynamic block to make it an optional configuration

## Motivation and Context
Instance refresh makes it such that we do not need an external entity to keep our ASGs updated.

Once such use case is a new AMI being released, causing a change in the EC2 Launch configuration attached to an ASG. Currently I've seen people have Cloudformation stacks to manage a "rolling" upgrade of the EC2 instances. There are other 3rd party tools that do this, however that is heavy weight especially since Instance Refresh has been released.

fixes https://github.com/terraform-aws-modules/terraform-aws-autoscaling/issues/130

## Breaking Changes
If you have any 3rd party tools that manage the EC2 instances within an ASG then this might cause issues in respect to multiple processes attempting to update the ASG. 

If this proves to be a problem, we can perhaps add a dynamic block mentioned previously to make this an optional configuration perhaps.

## How Has This Been Tested?
- [x] I have tested and validated these changes using one or more of the provided `examples/*` projects a successful plan has been generated:

<details><summary>Show Plan</summary>

```hcl
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

data.aws_vpc.default: Refreshing state...
data.aws_ami.amazon_linux: Refreshing state...
data.aws_subnet_ids.all: Refreshing state...
data.aws_security_group.default: Refreshing state...

------------------------------------------------------------------------

An execution plan has been generated and is shown below.
Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # aws_iam_service_linked_role.autoscaling will be created
  + resource "aws_iam_service_linked_role" "autoscaling" {
      + arn              = (known after apply)
      + aws_service_name = "autoscaling.amazonaws.com"
      + create_date      = (known after apply)
      + custom_suffix    = "something"
      + description      = "A service linked role for autoscaling"
      + id               = (known after apply)
      + name             = (known after apply)
      + path             = (known after apply)
      + unique_id        = (known after apply)
    }

  # module.example.aws_autoscaling_group.this[0] will be created
  + resource "aws_autoscaling_group" "this" {
      + arn                       = (known after apply)
      + availability_zones        = (known after apply)
      + default_cooldown          = 300
      + desired_capacity          = 1
      + enabled_metrics           = [
          + "GroupDesiredCapacity",
          + "GroupInServiceInstances",
          + "GroupMaxSize",
          + "GroupMinSize",
          + "GroupPendingInstances",
          + "GroupStandbyInstances",
          + "GroupTerminatingInstances",
          + "GroupTotalInstances",
        ]
      + force_delete              = false
      + health_check_grace_period = 300
      + health_check_type         = "EC2"
      + id                        = (known after apply)
      + launch_configuration      = (known after apply)
      + max_instance_lifetime     = 0
      + max_size                  = 1
      + metrics_granularity       = "1Minute"
      + min_elb_capacity          = 0
      + min_size                  = 0
      + name                      = (known after apply)
      + name_prefix               = (known after apply)
      + protect_from_scale_in     = false
      + service_linked_role_arn   = (known after apply)
      + termination_policies      = [
          + "Default",
        ]
      + vpc_zone_identifier       = [
          + "subnet-3f149576",
          + "subnet-8cd317d7",
          + "subnet-908e36f7",
        ]
      + wait_for_capacity_timeout = "0"

      + instance_refresh {
          + strategy = "Rolling"
          + triggers = [
              + "tag",
            ]

          + preferences {
              + min_healthy_percentage = 50
            }
        }

      + tag {
          + key                 = "Environment"
          + propagate_at_launch = true
          + value               = "dev"
        }
      + tag {
          + key                 = "Name"
          + propagate_at_launch = true
          + value               = "example-with-ec2"
        }
      + tag {
          + key                 = "Project"
          + propagate_at_launch = true
          + value               = "megasecret"
        }
      + tag {
          + key                 = "bar"
          + propagate_at_launch = true
        }
      + tag {
          + key                 = "extra_tag1"
          + propagate_at_launch = true
          + value               = "extra_value1"
        }
      + tag {
          + key                 = "extra_tag2"
          + propagate_at_launch = true
          + value               = "extra_value2"
        }
      + tag {
          + key                 = "foo"
          + propagate_at_launch = true
        }
    }

  # module.example.aws_launch_configuration.this[0] will be created
  + resource "aws_launch_configuration" "this" {
      + arn                         = (known after apply)
      + associate_public_ip_address = true
      + ebs_optimized               = false
      + enable_monitoring           = true
      + id                          = (known after apply)
      + image_id                    = "ami-0717dc2c13076628f"
      + instance_type               = "t2.micro"
      + key_name                    = (known after apply)
      + name                        = (known after apply)
      + name_prefix                 = "example-lc-"
      + placement_tenancy           = "default"
      + security_groups             = [
          + "sg-5782022f",
        ]
      + user_data_base64            = "IyEvYmluL2Jhc2gKZWNobyAiSGVsbG8gVGVycmFmb3JtISIK"

      + ebs_block_device {
          + delete_on_termination = true
          + device_name           = "/dev/xvdz"
          + encrypted             = (known after apply)
          + iops                  = (known after apply)
          + snapshot_id           = (known after apply)
          + volume_size           = 50
          + volume_type           = "gp2"
        }

      + metadata_options {
          + http_endpoint               = (known after apply)
          + http_put_response_hop_limit = (known after apply)
          + http_tokens                 = (known after apply)
        }

      + root_block_device {
          + delete_on_termination = true
          + encrypted             = (known after apply)
          + iops                  = (known after apply)
          + volume_size           = 50
          + volume_type           = "gp2"
        }
    }

  # module.example.null_resource.tags_as_list_of_maps[0] will be created
  + resource "null_resource" "tags_as_list_of_maps" {
      + id       = (known after apply)
      + triggers = {
          + "key"                 = "extra_tag1"
          + "propagate_at_launch" = "true"
          + "value"               = "extra_value1"
        }
    }

  # module.example.null_resource.tags_as_list_of_maps[1] will be created
  + resource "null_resource" "tags_as_list_of_maps" {
      + id       = (known after apply)
      + triggers = {
          + "key"                 = "extra_tag2"
          + "propagate_at_launch" = "true"
          + "value"               = "extra_value2"
        }
    }

  # module.example.random_pet.asg_name[0] will be created
  + resource "random_pet" "asg_name" {
      + id        = (known after apply)
      + keepers   = (known after apply)
      + length    = 2
      + separator = "-"
    }

Plan: 6 to add, 0 to change, 0 to destroy.

------------------------------------------------------------------------

Note: You didn't specify an "-out" parameter to save this plan, so Terraform
can't guarantee that exactly these actions will be performed if
"terraform apply" is subsequently run.
```

</details>

### CODE REVIEW

File: examples/asg_elb/main.tf [hunk 1]
```diff
@@ -101,7 +101,8 @@ module "example_asg" {
 # ELB
 ######
 module "elb" {
-  source = "terraform-aws-modules/elb/aws"
+  source  = "terraform-aws-modules/elb/aws"
+  version = "2.4.0"
```
Comment (spatel96) [inline examples/asg_elb/main.tf:105]: Pinning the ELB module to 2.4.0 to work around the provider source warning in the latest release causing the checks to fail.

File: main.tf [hunk 2]
```diff
@@ -96,6 +96,14 @@ resource "aws_autoscaling_group" "this" {
...
   max_instance_lifetime     = var.max_instance_lifetime
 
+  instance_refresh {
+    strategy = "Rolling"
+    preferences {
+      min_healthy_percentage = var.instance_refresh_minimum_health_percentage
+    }
+    triggers = var.instance_refresh_triggers
+  }
+
```
Comment (bobdoah) [inline main.tf:106]: Shouldn't this be a configuration option? This looks to be enabled by default.

**Review by bobdoah — CHANGES_REQUESTED** [review]

File: main.tf (hunk 2 above)
Comment (pmoranga) [reply main.tf:106]: I submitted a PR to the fork from @spatel96 adding a conditional flag to it.

Comment (spatel96) [reply main.tf]: Approved and merged, thanks!

**Review by bobdoah — APPROVED** [review]

**Review by pmoranga — APPROVED** [review]

File: variables.tf [hunk 3]
```diff
@@ -321,3 +321,20 @@ variable "max_instance_lifetime" {
   default     = 0
 }
 
+variable "instance_refresh_minimum_health_percentage" {
+  description = "The amount of capacity in the Auto Scaling group that must remain healthy during an instance refresh to allow the operation to continue, as a percentage of the desired capacity of the Auto Scaling group."
+  default     = 50
```
Comment (antonbabenko) [inline variables.tf:326]: Why `50`? [Magic number](https://en.wikipedia.org/wiki/Magic_number_(programming))?

File: variables.tf [hunk 4]
```diff
@@ -321,3 +321,20 @@ variable "max_instance_lifetime" {
   default     = 0
 }
 
+variable "instance_refresh_minimum_health_percentage" {
```
Comment (antonbabenko) [inline variables.tf:324]: Please remame this variable into `instance_refresh_min_healthy_percentage`

File: variables.tf (hunk 3 above)
Comment (antonbabenko) [reply variables.tf:326]: Make it default to `null` to use the provider's default which is `90`. Ref: https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/autoscaling_group#min_healthy_percentage

File: main.tf [hunk 5]
```diff
@@ -96,6 +96,17 @@ resource "aws_autoscaling_group" "this" {
   service_linked_role_arn   = var.service_linked_role_arn
   max_instance_lifetime     = var.max_instance_lifetime
 
+  dynamic "instance_refresh" {
+    for_each = var.instance_refresh_enabled ? [1] : []
+    content {
+      strategy = "Rolling"
+      preferences {
```
Comment (antonbabenko) [inline main.tf:103]: Please add [instance_warmup](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/autoscaling_group#instance_warmup) argument.

Comment (antonbabenko) [reply main.tf:103]: Also, block `preferences` should be `dynamic` since it is optional.

**Review by antonbabenko — CHANGES_REQUESTED** [review]
Hi!

This feature became available in [3.22.0](https://github.com/hashicorp/terraform-provider-aws/blob/main/CHANGELOG.md#3220-december-18-2020). Please update `versions.tf` accordingly.

File: variables.tf (hunk 3 above)
Comment (pmoranga) [reply variables.tf]: done! @spatel96 can u set as resolved?

File: variables.tf [hunk 6]
```diff
@@ -321,3 +321,20 @@ variable "max_instance_lifetime" {
...
+  description = "The amount of capacity in the Auto Scaling group that must remain healthy during an instance refresh to allow the operation to continue, as a percentage of the desired capacity of the Auto Scaling group."
+  default     = 50
+  type        = number
+}
+
+variable "instance_refresh_triggers" {
+  description = "Set of additional property names that will trigger an Instance Refresh. A refresh will always be triggered by a change in any of launch_configuration, launch_template, or mixed_instances_policy"
+  default     = ["tag"]
```
[the remark this answers is not in the document]
Comment (pmoranga) [reply variables.tf]: done! Added the usage of the tag as trigger in the README.md example! @spatel96 can u set as resolved?

File: variables.tf [hunk 7]
```diff
@@ -321,3 +321,20 @@ variable "max_instance_lifetime" {
...
+  type        = list(string)
+}
+
+variable "instance_refresh_enabled" {
+  description = "If true an Instance Refresh will start when the Auto Scaling Group is updated"
+  default     = false
```
[the remark this answers is not in the document]
Comment (pmoranga) [reply variables.tf]: done! @spatel96 can you set as resolved?

File: variables.tf (hunk 4 above)
Comment (pmoranga) [reply variables.tf]: done! @spatel96 can you set as resolved?

File: main.tf (hunk 5 above)
Comment (pmoranga) [reply main.tf]: done! @spatel96 can you set as resolved?

**Pull request closed without merging**